JavaScript heap out of memory with CRA based React/TypeScript application

**Summary**

Running Stryker on a Create React App base TypeScript project causes an FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

**Performance profile of the application before the fatal error**

![image](https://user-images.githubusercontent.com/3197730/66814681-4e745f80-ef37-11e9-9752-9b627aa0838e.png)

**Stryker config**

```js
// This config was generated using a preset.
// Please see the handbook for more information: https://github.com/stryker-mutator/stryker-handbook/blob/master/stryker/guides/react.md#react
module.exports = function(config) {
  config.set({
    mutate: ["src/**/*.ts?(x)", "!src/**/*@(.test|.spec|Spec).ts?(x)"],
    mutator: "typescript",
    testRunner: "jest",
    reporters: ["progress", "clear-text", "html"],
    tempDirName: "stryker-tmp",
    coverageAnalysis: "off",
    jest: {
      projectType: "react"
    }
  });
};
```

**Test runner config**

```
  "jest": {
    "globalSetup": "./src/setupTests.ts",
    "coverageThreshold": {
      "./src/**/*.{ts,tsx}": {
        "lines": 80,
        "statements": 80,
        "branches": 80,
        "functions": 80
      }
    },
    "coverageReporters": [
      "cobertura"
    ]
  },

```

setupTests.ts
```
export default (): void => {
  (global as any).alert = (): void => {};
};
```

**Stryker environment**

<!-- Please list your stryker plugins + versions here (output of `npm ls | grep stryker`) -->

```
+-- @stryker-mutator/core@2.1.0
| +-- @stryker-mutator/api@2.1.0
| +-- @stryker-mutator/util@2.1.0
+-- @stryker-mutator/html-reporter@2.1.0
| +-- @stryker-mutator/api@2.1.0 deduped
| +-- @stryker-mutator/util@2.1.0 deduped
+-- @stryker-mutator/jest-runner@2.1.0
| +-- @stryker-mutator/api@2.1.0 deduped
+-- @stryker-mutator/typescript@2.1.0
| +-- @stryker-mutator/api@2.1.0 deduped
| +-- @stryker-mutator/util@2.1.0 deduped
```

<!-- Please also add the test runner you are using.  Examples:-->

```
jest@24.9.0
@testing-library/react@9.1.4
jsdom@14.1.0
full-icu@1.3.0
```

**Test runner environment**

<!-- Please add your test command here (probably the command you use in `npm test`) command -->

```shell
cross-env NODE_ICU_DATA=node_modules/full-icu react-scripts test
```

<!-- Please add any config files that are used by your test runner. For example jest.config.js, karma.conf.js, angular.json, mocha.opts, ...  -->

```json
```


**Your Environment**

| software         | version(s)
| ---------------- | -------  
| node             | v10.15.3
| npm              | 6.11.3
| Operating System | Windows 10 version 1809

**Add stryker.log**
08:38:37 (5104) INFO ConfigReader Using stryker.conf.js in the current working directory.
08:38:37 (5104) DEBUG BroadcastReporter Broadcasting to reporters ["progress","clear-text","html"]
08:38:37 (5104) DEBUG ConfigReader Loading config stryker.conf.js
08:38:37 (5104) DEBUG PluginLoader Loading @stryker-mutator/* from C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator
08:38:37 (5104) DEBUG PluginLoader Loading plugin "C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\html-reporter" (matched with expression @stryker-mutator/*)
08:38:37 (5104) DEBUG PluginLoader Loading plugin "C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\jest-runner" (matched with expression @stryker-mutator/*)
08:38:37 (5104) DEBUG PluginLoader Loading plugin "C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\typescript" (matched with expression @stryker-mutator/*)
08:38:37 (5104) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\html-reporter
08:38:37 (5104) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\jest-runner
08:38:37 (5104) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\typescript
08:38:37 (5104) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\core\src\reporters\index.js
08:38:37 (5104) DEBUG TypescriptConfigEditor No 'tsconfigFile' specified, not loading any config
08:38:38 (5104) INFO InputFileResolver Found 516 of 1078 file(s) to be mutated.
08:38:38 (5104) DEBUG InputFileResolver All input files: [
  "C:\\Repos\\Adviesbox-Advies-React\\.azuredevops\\pull_request_template.md",
  "C:\\Repos\\Adviesbox-Advies-React\\.eslintignore",
  "C:\\Repos\\Adviesbox-Advies-React\\.eslintrc.js",
  "C:\\Repos\\Adviesbox-Advies-React\\.gitignore",
  "C:\\Repos\\Adviesbox-Advies-React\\.prettierrc",
  "C:\\Repos\\Adviesbox-Advies-React\\.vscode\\launch.json",
  "C:\\Repos\\Adviesbox-Advies-React\\README.md",
  ... many more omitted files
  "C:\\Repos\\Adviesbox-Advies-React\\src\\zoekscherm\\zoekscherm.tsx"
]
08:52:42 (7064) DEBUG TemporaryDirectory Using temp directory "C:\Repos\Adviesbox-Advies-React\stryker-tmp"
08:52:42 (7064) DEBUG TestFrameworkOrchestrator The `coverageAnalysis` setting is "off", not hooking into the test framework to achieve performance benefits.
08:52:42 (7064) INFO InitialTestExecutor Starting initial test run. This may take a while.
08:52:42 (7064) DEBUG Sandbox Creating a sandbox for files in C:\Repos\Adviesbox-Advies-React\stryker-tmp\sandbox9953131
08:52:43 (7064) DEBUG Sandbox Creating test runner 0
08:52:43 (7064) DEBUG ChildProcessProxy Starting C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\core\src\test-runner\ChildProcessTestRunnerWorker.js in child process 19376
08:52:43 (19376) DEBUG ChildProcessProxyWorker Changing current working directory for this process to C:\Repos\Adviesbox-Advies-React\stryker-tmp\sandbox9953131
08:52:43 (19376) DEBUG PluginLoader Loading @stryker-mutator/* from C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator
08:52:43 (19376) DEBUG PluginLoader Loading plugin "C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\html-reporter" (matched with expression @stryker-mutator/*)
08:52:43 (19376) DEBUG PluginLoader Loading plugin "C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\jest-runner" (matched with expression @stryker-mutator/*)
08:52:43 (19376) DEBUG PluginLoader Loading plugin "C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\typescript" (matched with expression @stryker-mutator/*)
08:52:43 (19376) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\html-reporter
08:52:43 (19376) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\jest-runner
08:52:43 (19376) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\typescript
08:52:43 (19376) DEBUG PluginLoader Loading plugins C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\core\src\reporters\index.js
08:52:44 (19376) DEBUG JestTestRunner Running jest with --findRelatedTests flag. Set jest.enableFindRelatedTests to false to run all tests on every mutant.
08:52:44 (19376) DEBUG JestTestRunner Project root is C:\Repos\Adviesbox-Advies-React\stryker-tmp\sandbox9953131
08:52:44 (7064) DEBUG TimeoutDecorator Starting timeout timer (300000 ms) for a test run
08:52:44 (19376) TRACE JestPromiseTestAdapter Invoking Jest with config {"roots":["<rootDir>/src"],"collectCoverageFrom":["src/**/*.{js,jsx,ts,tsx}","!src/**/*.d.ts"],"setupFiles":["C:\\Repos\\Adviesbox-Advies-React\\node_modules\\react-app-polyfill\\jsdom.js"],"setupFilesAfterEnv":["<rootDir>/src/setupTests.ts"],"testMatch":["<rootDir>/src/**/__tests__/**/*.{js,jsx,ts,tsx}","<rootDir>/src/**/*.{spec,test}.{js,jsx,ts,tsx}"],"testEnvironment":"jsdom","transform":{"^.+\\.(js|jsx|ts|tsx)$":"C:\\Repos\\Adviesbox-Advies-React\\node_modules\\react-scripts\\config\\jest\\babelTransform.js","^.+\\.css$":"C:\\Repos\\Adviesbox-Advies-React\\node_modules\\react-scripts\\config\\jest\\cssTransform.js","^(?!.*\\.(js|jsx|ts|tsx|css|json)$)":"C:\\Repos\\Adviesbox-Advies-React\\node_modules\\react-scripts\\config\\jest\\fileTransform.js"},"transformIgnorePatterns":["[/\\\\]node_modules[/\\\\].+\\.(js|jsx|ts|tsx)$","^.+\\.module\\.(css|sass|scss)$"],"modulePaths":[],"moduleNameMapper":{"^react-native$":"react-native-web","^.+\\.module\\.(css|sass|scss)$":"identity-obj-proxy"},"moduleFileExtensions":["web.js","js","web.ts","ts","web.tsx","tsx","json","web.jsx","jsx","node"],"watchPlugins":["jest-watch-typeahead/filename","jest-watch-typeahead/testname"],"rootDir":"C:\\Repos\\Adviesbox-Advies-React\\stryker-tmp\\sandbox9953131","coverageReporters":["cobertura"],"coverageThreshold":{"./src/**/*.{ts,tsx}":{"lines":80,"statements":80,"branches":80,"functions":80}},"globalSetup":"./src/setupTests.ts","collectCoverage":false,"verbose":false,"bail":false,"notify":false,"reporters":[]}
08:55:41 (7064) DEBUG ChildProcessProxy Disposing of worker process 19376
08:55:41 (7064) DEBUG ChildProcessProxy Kill 19376
08:55:41 (7064) INFO InitialTestExecutor Initial test run succeeded. Ran 1531 tests in 2 minutes 59 seconds (net 72308 ms, overhead 104839 ms).

<!-- Please add your stryker.log file. This file can be generated using `stryker run --fileLogLevel trace`. You can drag and drop it here. -->
<!-- Your source code is never logged to this file, however file names are logged. Feel free to obfuscate those log messages if you think it is a problem -->

@simondel Yes I have seen `MutatorFacade 26902 Mutant(s) generated`.

Seems the last part of the output is missing from the log file. Guess a result of the crash. Still had a copy:

```
08:52:44 (7064) DEBUG TimeoutDecorator Starting timeout timer (300000 ms) for a test run
08:55:41 (7064) DEBUG ChildProcessProxy Disposing of worker process 19376
08:55:41 (7064) DEBUG ChildProcessProxy Kill 19376
08:55:41 (7064) INFO InitialTestExecutor Initial test run succeeded. Ran 1531 tests in 2 minutes 59 seconds (net 72308 ms, overhead 104839 ms).
08:55:43 (7064) INFO MutatorFacade 26902 Mutant(s) generated

<--- Last few GCs --->

[7064:000002BECD70BE50]   196676 ms: Scavenge 1396.2 (1424.7) -> 1395.7 (1425.2) MB, 2.2 / 0.0 ms  (average mu = 0.265, current mu = 0.188) allocation failure [7064:000002BECD70BE50]   197712 ms: Mark-sweep 1396.9 (1425.2) -> 1396.6 (1425.7) MB, 858.0 / 0.0 ms  (+ 0.0 ms in 23 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 1040 ms) (average mu = 0.229, current mu = 0.185) all

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 000003B2BE65C5C1]
Security context: 0x011fc951e6e9 <JSObject>
    1: /* anonymous */(aka /* anonymous */) [000002E91AC02201] [C:\Repos\Adviesbox-Advies-React\node_modules\@stryker-mutator\core\src\mutants\MutantTestMatcher.js:~29] [pc=000003B2BE97E4CC](this=0x013e08c026f1 <undefined>,mutant=0x02e91ac049b9 <TestableMutant map = 0000026C48CCA191>)
    2: arguments adaptor frame: 3->1
    3: forEach [0000011FC9506DF1](t...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF735F4F04A v8::internal::GCIdleTimeHandler::GCIdleTimeHandler+5114
 2: 00007FF735F2A0C6 node::MakeCallback+4518
 3: 00007FF735F2AA30 node_module_register+2032
 4: 00007FF7361B20EE v8::internal::FatalProcessOutOfMemory+846
 5: 00007FF7361B201F v8::internal::FatalProcessOutOfMemory+639
 6: 00007FF7366D2BC4 v8::internal::Heap::MaxHeapGrowingFactor+9556
 7: 00007FF7366C9C46 v8::internal::ScavengeJob::operator=+24310
 8: 00007FF7366C829C v8::internal::ScavengeJob::operator=+17740
 9: 00007FF7366D0F87 v8::internal::Heap::MaxHeapGrowingFactor+2327
10: 00007FF7366D1006 v8::internal::Heap::MaxHeapGrowingFactor+2454
11: 00007FF73628CDB7 v8::internal::Factory::NewFillerObject+55
12: 00007FF736323179 v8::internal::WasmJs::Install+30617
13: 000003B2BE65C5C1
npm ERR! code ELIFECYCLE
npm ERR! errno 134
npm ERR! adviesbox@0.1.0 stryker: `cross-env NODE_ICU_DATA=node_modules/full-icu stryker run "--fileLogLevel" "trace" "--logLevel" "debug"`
npm ERR! Exit status 134
npm ERR!
npm ERR! Failed at the adviesbox@0.1.0 stryker script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\mauri\AppData\Roaming\npm-cache\_logs\2019-10-15T06_55_59_764Z-debug.log
```

--- a/packages/api/src/report/MatchedMutant.ts
+++ b/packages/api/src/report/MatchedMutant.ts
@@ -1,9 +1,31 @@
 interface MatchedMutant {
+  /**
+   * The ID of the mutant
+   */
   readonly id: string;
+  /**
+   * The Mutator name
+   */
   readonly mutatorName: string;
+  /**
+   * Whether or not all tests will run for this mutant
+   */
+  readonly runAllTests: boolean;
+  /**
+   * If not all tests will run for this mutant, this array will contain the ids of the tests that will run.
+   */
   readonly scopedTestIds: number[];
+  /**
+   * The time spent on the tests that will run in initial test run
+   */
   readonly timeSpentScopedTests: number;
+  /**
+   * The file name that contains the mutant
+   */
   readonly fileName: string;
+  /**
+   * The replacement code to be inserted
+   */
   readonly replacement: string;
 }
 

--- a/packages/core/src/Sandbox.ts
+++ b/packages/core/src/Sandbox.ts
@@ -94,13 +94,13 @@ export default class Sandbox {
           `Transpile error occurred: "${transpiledMutant.transpileResult.error}" during transpiling of mutant ${transpiledMutant.mutant.toString()}`
         );
       }
-      const result = transpiledMutant.mutant.result(MutantStatus.TranspileError, []);
+      const result = transpiledMutant.mutant.createResult(MutantStatus.TranspileError, []);
       return result;
     } else if (!transpiledMutant.mutant.selectedTests.length) {
-      const result = transpiledMutant.mutant.result(MutantStatus.NoCoverage, []);
+      const result = transpiledMutant.mutant.createResult(MutantStatus.NoCoverage, []);
       return result;
     } else if (!transpiledMutant.changedAnyTranspiledFiles) {
-      const result = transpiledMutant.mutant.result(MutantStatus.Survived, []);
+      const result = transpiledMutant.mutant.createResult(MutantStatus.Survived, []);
       return result;
     } else {
       // No early result possible, need to run in the sandbox later
@@ -115,7 +115,7 @@ export default class Sandbox {
       const error = runResult.errorMessages ? runResult.errorMessages.toString() : '(undefined)';
       this.log.debug('A runtime error occurred: %s during execution of mutant: %s', error, mutant.toString());
     }
-    return mutant.result(status, testNames);
+    return mutant.createResult(status, testNames);
   }
 
   private determineMutantState(runResult: RunResult): MutantStatus {
@@ -195,7 +195,7 @@ export default class Sandbox {
   }
 
   private getFilterTestsHooks(mutant: TestableMutant): string | undefined {
-    if (this.testFramework) {
+    if (this.testFramework && !mutant.runAllTests) {
       return wrapInClosure(this.testFramework.filter(mutant.selectedTests));
     } else {
       return undefined;

--- a/packages/core/src/Sandbox.ts
+++ b/packages/core/src/Sandbox.ts
@@ -96,7 +96,7 @@ export default class Sandbox {
       }
       const result = transpiledMutant.mutant.createResult(MutantStatus.TranspileError, []);
       return result;
-    } else if (!transpiledMutant.mutant.selectedTests.length) {
+    } else if (!transpiledMutant.mutant.runAllTests && !transpiledMutant.mutant.selectedTests.length) {
       const result = transpiledMutant.mutant.createResult(MutantStatus.NoCoverage, []);
       return result;
     } else if (!transpiledMutant.changedAnyTranspiledFiles) {

--- a/packages/core/src/TestableMutant.ts
+++ b/packages/core/src/TestableMutant.ts
@@ -18,7 +18,7 @@ class TestFilter {
   public selectedTests: TestSelection[] = [];
 
   public selectAllTests(runResult: RunResult) {
-    runResult.tests.forEach(testResult => (this.timeSpentScopedTests += testResult.timeSpentMs));
+    this.timeSpentScopedTests = runResult.tests.reduce((time, test) => time + test.timeSpentMs, this.timeSpentScopedTests);
     this.runAllTests = true;
   }
 

--- a/packages/core/src/TestableMutant.ts
+++ b/packages/core/src/TestableMutant.ts
@@ -12,19 +12,39 @@ export enum TestSelectionResult {
   Success
 }
 
+class TestFilter {
+  public timeSpentScopedTests = 0;
+  public runAllTests = false;
+  public selectedTests: TestSelection[] = [];
+
+  public selectAllTests(runResult: RunResult) {
+    runResult.tests.forEach(testResult => (this.timeSpentScopedTests += testResult.timeSpentMs));
+    this.runAllTests = true;
+  }
+
+  public selectTest(testResult: TestResult, index: number) {
+    this.selectedTests.push({ id: index, name: testResult.name });
+    this.timeSpentScopedTests += testResult.timeSpentMs;
+    this.runAllTests = false;
+  }
+}
+
 export default class TestableMutant {
-  private readonly _selectedTests: TestSelection[] = [];
   public specsRan: string[] = [];
-  private _timeSpentScopedTests = 0;
+  private readonly filter = new TestFilter();
   private _location: Location;
   public testSelectionResult = TestSelectionResult.Success;
 
   public get selectedTests(): TestSelection[] {
-    return this._selectedTests;
+    return this.filter.selectedTests;
+  }
+
+  public get runAllTests(): boolean {
+    return this.filter.runAllTests;
   }
 
   public get timeSpentScopedTests() {
-    return this._timeSpentScopedTests;
+    return this.filter.timeSpentScopedTests;
   }
 
   public get fileName() {
@@ -59,13 +79,12 @@ export default class TestableMutant {
   }
 
   public selectAllTests(runResult: RunResult, testSelectionResult: TestSelectionResult) {
+    this.filter.selectAllTests(runResult);
     this.testSelectionResult = testSelectionResult;
-    runResult.tests.forEach((testResult, id) => this.selectTest(testResult, id));
   }
 
   public selectTest(testResult: TestResult, index: number) {
-    this._selectedTests.push({ id: index, name: testResult.name });
-    this._timeSpentScopedTests += testResult.timeSpentMs;
+    this.filter.selectTest(testResult, index);
   }
 
   constructor(public readonly id: string, public mutant: Mutant, public sourceFile: SourceFile) {}
@@ -96,7 +115,7 @@ export default class TestableMutant {
     return [startIndexLines, endIndexLines];
   }
 
-  public result(status: MutantStatus, testsRan: string[]): MutantResult {
+  public createResult(status: MutantStatus, testsRan: string[]): MutantResult {
     return freezeRecursively({
       id: this.id,
       location: this.location,

--- a/packages/core/src/mutants/MutantTestMatcher.ts
+++ b/packages/core/src/mutants/MutantTestMatcher.ts
@@ -4,7 +4,6 @@ import { Mutant } from '@stryker-mutator/api/mutant';
 import { commonTokens, tokens } from '@stryker-mutator/api/plugin';
 import { MatchedMutant } from '@stryker-mutator/api/report';
 import { CoverageCollection, CoveragePerTestResult, CoverageResult, StatementMap } from '@stryker-mutator/api/test_runner';
-import * as _ from 'lodash';
 import { coreTokens } from '../di';
 import InputFileCollection from '../input/InputFileCollection';
 import { InitialTestRunResult } from '../process/InitialTestExecutor';
@@ -143,14 +142,15 @@ export class MutantTestMatcher {
    * @returns The MatchedMutant
    */
   private mapMutantOnMatchedMutant(testableMutant: TestableMutant): MatchedMutant {
-    const matchedMutant = _.cloneDeep({
+    const matchedMutant: MatchedMutant = {
       fileName: testableMutant.mutant.fileName,
       id: testableMutant.id,
       mutatorName: testableMutant.mutant.mutatorName,
       replacement: testableMutant.mutant.replacement,
+      runAllTests: testableMutant.runAllTests,
       scopedTestIds: testableMutant.selectedTests.map(testSelection => testSelection.id),
       timeSpentScopedTests: testableMutant.timeSpentScopedTests
-    });
+    };
     return Object.freeze(matchedMutant);
   }
 

--- a/packages/core/src/reporters/ProgressKeeper.ts
+++ b/packages/core/src/reporters/ProgressKeeper.ts
@@ -14,7 +14,7 @@ abstract class ProgressKeeper implements Reporter {
 
   public onAllMutantsMatchedWithTests(matchedMutants: readonly MatchedMutant[]): void {
     this.timer = new Timer();
-    this.mutantIdsWithoutCoverage = matchedMutants.filter(m => m.scopedTestIds.length === 0).map(m => m.id);
+    this.mutantIdsWithoutCoverage = matchedMutants.filter(m => !m.runAllTests && m.scopedTestIds.length === 0).map(m => m.id);
     this.progress.total = matchedMutants.length - this.mutantIdsWithoutCoverage.length;
   }
 

--- a/packages/core/src/utils/objectUtils.ts
+++ b/packages/core/src/utils/objectUtils.ts
@@ -16,14 +16,6 @@ export function filterEmpty<T>(input: Array<T | null | void>) {
   return input.filter(item => item !== undefined && item !== null) as T[];
 }
 
-export function copy<T>(obj: T, deep?: boolean) {
-  if (deep) {
-    return _.cloneDeep(obj);
-  } else {
-    return _.clone(obj);
-  }
-}
-
 export function wrapInClosure(codeFragment: string) {
   return `
     (function (window) {

--- a/packages/core/test/unit/Sandbox.spec.ts
+++ b/packages/core/test/unit/Sandbox.spec.ts
@@ -258,6 +258,14 @@ describe(Sandbox.name, () => {
       expect(fileUtils.writeFile).not.calledWith(expectedTestFrameworkHooksFile);
     });
 
+    it('should not filter any tests when runAllTests = true', async () => {
+      const sut = await createSut();
+      const mutant = new TestableMutant('2', createMutant(), new SourceFile(new File('', '')));
+      mutant.selectAllTests(runResult, TestSelectionResult.Failed);
+      sut.runMutant(new TranspiledMutant(mutant, { outputFiles: [new File(expectedTargetFileToMutate, '')], error: null }, true));
+      expect(fileUtils.writeFile).not.calledWith(expectedTestFrameworkHooksFile);
+    });
+
     it('should report a runtime error when test run errored', async () => {
       runResult.status = RunStatus.Error;
       runResult.errorMessages = ['Cannot call "foo" of undefined (or something)'];

--- a/packages/core/test/unit/Sandbox.spec.ts
+++ b/packages/core/test/unit/Sandbox.spec.ts
@@ -254,8 +254,22 @@ describe(Sandbox.name, () => {
     it('should not filter any tests when testFramework = null', async () => {
       const sut = await createSut();
       const mutant = new TestableMutant('2', createMutant(), new SourceFile(new File('', '')));
-      sut.runMutant(new TranspiledMutant(mutant, { outputFiles: [new File(expectedTargetFileToMutate, '')], error: null }, true));
+      await sut.runMutant(new TranspiledMutant(mutant, { outputFiles: [new File(expectedTargetFileToMutate, '')], error: null }, true));
       expect(fileUtils.writeFile).not.calledWith(expectedTestFrameworkHooksFile);
+    });
+
+    it('should not filter any tests when runAllTests = true', async () => {
+      // Arrange
+      while (transpiledMutant.mutant.selectedTests.pop());
+      transpiledMutant.mutant.selectAllTests(runResult, TestSelectionResult.Success);
+      const sut = await createSut();
+
+      // Act
+      await sut.runMutant(transpiledMutant);
+
+      // Assert
+      expect(fileUtils.writeFile).not.calledWith(expectedTestFrameworkHooksFile);
+      expect(testRunner.run).called;
     });
 
     it('should not filter any tests when runAllTests = true', async () => {
@@ -287,6 +301,18 @@ describe(Sandbox.name, () => {
         `Transpile error occurred: "Error! Cannot negate a string (or something)" during transpiling of mutant ${transpiledMutant.mutant.toString()}`
       );
       expect(mutantResult.status).eq(MutantStatus.TranspileError);
+    });
+
+    it('should report an early result when there are no files scoped', async () => {
+      // Arrange
+      while (transpiledMutant.mutant.selectedTests.pop());
+
+      // Act
+      const sut = await createSut();
+
+      // Assert
+      const mutantResult = await sut.runMutant(transpiledMutant);
+      expect(mutantResult.status).eq(MutantStatus.NoCoverage);
     });
 
     it('should report an early result when there are no file changes', async () => {

--- a/packages/core/test/unit/TestableMutant.spec.ts
+++ b/packages/core/test/unit/TestableMutant.spec.ts
@@ -5,7 +5,7 @@ import { expect } from 'chai';
 import SourceFile from '../../src/SourceFile';
 import TestableMutant, { TestSelectionResult } from '../../src/TestableMutant';
 
-describe('TestableMutant', () => {
+describe(TestableMutant.name, () => {
   let innerMutant: Mutant;
 
   beforeEach(() => {
@@ -29,8 +29,17 @@ describe('TestableMutant', () => {
       TestSelectionResult.FailedButAlreadyReported
     );
     expect(sut.timeSpentScopedTests).eq(54);
-    expect(sut.selectedTests).deep.eq([{ id: 0, name: 'spec1' }, { id: 1, name: 'spec2' }]);
+    expect(sut.runAllTests).true;
     expect(sut.testSelectionResult).eq(TestSelectionResult.FailedButAlreadyReported);
+  });
+
+  it('should scope tests when selecting individual tests', () => {
+    const sut = new TestableMutant('3', innerMutant, new SourceFile(new File('foobar.js', 'alert("foobar")')));
+    sut.selectTest(testResult({ name: 'spec1', timeSpentMs: 12 }), 0);
+    sut.selectTest(testResult({ name: 'spec3', timeSpentMs: 32 }), 2);
+    expect(sut.timeSpentScopedTests).eq(44);
+    expect(sut.selectedTests).deep.eq([{ id: 0, name: 'spec1' }, { id: 2, name: 'spec3' }]);
+    expect(sut.testSelectionResult).eq(TestSelectionResult.Success);
   });
 
   it('should calculate position using sourceFile', () => {

--- a/packages/core/test/unit/mutants/MutantTestMatcher.spec.ts
+++ b/packages/core/test/unit/mutants/MutantTestMatcher.spec.ts
@@ -89,13 +89,11 @@ describe(MutantTestMatcher.name, () => {
         });
 
         describe('without code coverage info', () => {
-          it('should add both tests to the mutants and report failure', async () => {
-            const expectedTestSelection = [{ id: 0, name: 'test one' }, { id: 1, name: 'test two' }];
-
+          it('should run all tests for the mutants and report failure', async () => {
             const result = await sut.matchWithMutants(mutants);
 
-            expect(result[0].selectedTests).deep.eq(expectedTestSelection);
-            expect(result[1].selectedTests).deep.eq(expectedTestSelection);
+            expect(result[0].runAllTests).true;
+            expect(result[1].runAllTests).true;
             expect(TestSelectionResult[result[0].testSelectionResult]).eq(TestSelectionResult[TestSelectionResult.FailedButAlreadyReported]);
             expect(TestSelectionResult[result[1].testSelectionResult]).eq(TestSelectionResult[TestSelectionResult.FailedButAlreadyReported]);
             expect(testInjector.logger.warn).calledWith(
@@ -113,7 +111,8 @@ describe(MutantTestMatcher.name, () => {
                 id: '0',
                 mutatorName: result[0].mutatorName,
                 replacement: result[0].replacement,
-                scopedTestIds: result[0].selectedTests.map(test => test.id),
+                runAllTests: true,
+                scopedTestIds: [],
                 timeSpentScopedTests: result[0].timeSpentScopedTests
               },
               {
@@ -121,7 +120,8 @@ describe(MutantTestMatcher.name, () => {
                 id: '1',
                 mutatorName: result[1].mutatorName,
                 replacement: result[1].replacement,
-                scopedTestIds: result[1].selectedTests.map(test => test.id),
+                runAllTests: true,
+                scopedTestIds: [],
                 timeSpentScopedTests: result[1].timeSpentScopedTests
               }
             ];
@@ -254,12 +254,10 @@ describe(MutantTestMatcher.name, () => {
           });
 
           it('should select all test in the test run but not report the error yet', async () => {
-            const expectedTestSelection: TestSelection[] = [{ name: 'test one', id: 0 }, { name: 'test two', id: 1 }];
-
             const result = await sut.matchWithMutants(mutants);
 
-            expect(result[0].selectedTests).deep.eq(expectedTestSelection);
-            expect(result[1].selectedTests).deep.eq(expectedTestSelection);
+            expect(result[0].runAllTests).true;
+            expect(result[1].runAllTests).true;
             expect(result[0].testSelectionResult).eq(TestSelectionResult.Failed);
             expect(result[1].testSelectionResult).eq(TestSelectionResult.Failed);
             expect(testInjector.logger.warn).not.called;
@@ -315,12 +313,10 @@ describe(MutantTestMatcher.name, () => {
           });
 
           it('should add all test results to the mutant that is covered by the baseline', async () => {
-            const expectedTestSelection = [{ id: 0, name: 'test one' }, { id: 1, name: 'test two' }];
-
             const result = await sut.matchWithMutants(mutants);
 
-            expect(result[0].selectedTests).deep.eq(expectedTestSelection);
-            expect(result[1].selectedTests).deep.eq(expectedTestSelection);
+            expect(result[0].runAllTests).true;
+            expect(result[1].runAllTests).true;
           });
         });
       });
@@ -458,12 +454,11 @@ describe(MutantTestMatcher.name, () => {
     it('should match all mutants to all tests and log a warning when there is no coverage data', async () => {
       mutants.push(mutant({ fileName: 'fileWithMutantOne' }), mutant({ fileName: 'fileWithMutantTwo' }));
       initialRunResult.runResult.tests.push(testResult(), testResult());
-      const expectedTestSelection: TestSelection[] = [{ id: 0, name: 'name' }, { id: 1, name: 'name' }];
 
       const result = await sut.matchWithMutants(mutants);
 
-      expect(result[0].selectedTests).deep.eq(expectedTestSelection);
-      expect(result[1].selectedTests).deep.eq(expectedTestSelection);
+      expect(result[0].runAllTests).true;
+      expect(result[1].runAllTests).true;
       expect(result[0].testSelectionResult).deep.eq(TestSelectionResult.FailedButAlreadyReported);
       expect(result[1].testSelectionResult).deep.eq(TestSelectionResult.FailedButAlreadyReported);
       expect(testInjector.logger.warn).to.have.been.calledWith(
@@ -537,12 +532,11 @@ describe(MutantTestMatcher.name, () => {
     it('should match all mutants to all tests', async () => {
       mutants.push(mutant({ fileName: 'fileWithMutantOne' }), mutant({ fileName: 'fileWithMutantTwo' }));
       initialRunResult.runResult.tests.push(testResult(), testResult());
-      const expectedTestSelection = [{ id: 0, name: 'name' }, { id: 1, name: 'name' }];
 
       const result = await sut.matchWithMutants(mutants);
 
-      expect(result[0].selectedTests).deep.eq(expectedTestSelection);
-      expect(result[1].selectedTests).deep.eq(expectedTestSelection);
+      expect(result[0].runAllTests).true;
+      expect(result[1].runAllTests).true;
     });
   });
 

--- a/packages/core/test/unit/reporters/ProgressReporter.spec.ts
+++ b/packages/core/test/unit/reporters/ProgressReporter.spec.ts
@@ -57,7 +57,7 @@ describe('ProgressReporter', () => {
 
         sut.onAllMutantsMatchedWithTests(matchedMutants);
       });
-      
+
       it('the total of MatchedMutants in the progress bar should be 2', () => {
         expect(progressBarModule.default).to.have.been.calledWithMatch(progressBarContent, { total: 2 });
       });

--- a/packages/core/test/unit/reporters/ProgressReporter.spec.ts
+++ b/packages/core/test/unit/reporters/ProgressReporter.spec.ts
@@ -51,6 +51,17 @@ describe('ProgressReporter', () => {
         expect(progressBarModule.default).to.have.been.calledWithMatch(progressBarContent, { total: 2 });
       });
     });
+    describe('when mutants match to all tests', () => {
+      beforeEach(() => {
+        matchedMutants = [matchedMutant(0, '0', true), matchedMutant(0, '1', true)];
+
+        sut.onAllMutantsMatchedWithTests(matchedMutants);
+      });
+      
+      it('the total of MatchedMutants in the progress bar should be 2', () => {
+        expect(progressBarModule.default).to.have.been.calledWithMatch(progressBarContent, { total: 2 });
+      });
+    });
   });
 
   describe('onMutantTested()', () => {

--- a/packages/core/test/unit/utils/TemporaryDirectory.spec.ts
+++ b/packages/core/test/unit/utils/TemporaryDirectory.spec.ts
@@ -55,7 +55,7 @@ describe(TemporaryDirectory.name, () => {
   describe('dispose', () => {
     describe('when temp directory is initialized', () => {
       beforeEach(() => sut.initialize());
-      it('should call deleteDir fileApi', () => {
+      it('should call deleteDir fileApi', async () => {
         const expectedPath = path.resolve(tempDirName);
         deleteDirStub.resolves('delResolveStub');
 

--- a/packages/test-helpers/src/factory.ts
+++ b/packages/test-helpers/src/factory.ts
@@ -184,7 +184,7 @@ export function transpiler(): sinon.SinonStubbedInstance<Transpiler> {
   };
 }
 
-export function matchedMutant(numberOfTests: number, mutantId = numberOfTests.toString()): MatchedMutant {
+export function matchedMutant(numberOfTests: number, mutantId = numberOfTests.toString(), runAllTests = false): MatchedMutant {
   const scopedTestIds: number[] = [];
   for (let i = 0; i < numberOfTests; i++) {
     scopedTestIds.push(1);
@@ -193,6 +193,7 @@ export function matchedMutant(numberOfTests: number, mutantId = numberOfTests.to
     fileName: '',
     id: mutantId,
     mutatorName: '',
+    runAllTests,
     replacement: '',
     scopedTestIds,
     timeSpentScopedTests: 0

--- a/packages/wct-runner/test/helpers/initChai.ts
+++ b/packages/wct-runner/test/helpers/initChai.ts
@@ -1,6 +1,5 @@
 import * as chai from 'chai';
 import * as chaiAsPromised from 'chai-as-promised';
 import * as sinonChai from 'sinon-chai';
-
-chai.use(chaiAsPromised);
 chai.use(sinonChai);
+chai.use(chaiAsPromised);

